### ISSUE TITLE
GELF Kafka Input - New consumer API (non-legacy) mode

### ISSUE THREAD

**Issue by samupr** [Graylog2/graylog2-server#9927]
## Expected Behavior
Kafka logs should get fetched without any errors.

## Current Behavior
Graylog log is continuously full of these errors:-
2021-01-18T22:54:15.192Z ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
2021-01-18T22:54:15.192Z ERROR [KafkaTransport] Caught recoverable exception. Retrying

## Context
I switched from the legacy mode (connection to zookeeper)

## Your Environment
cluster of 6 kafka servers (kafka-2.7.0-1)
centos 7

* Graylog Version: graylog-server-4.0.1-1
* Java Version: java-1.8.0
* Elasticsearch Version:  elasticsearch-oss-7.10.1-1
* MongoDB Version: mongodb-org-4.0.20-1
* Operating System: Centos 7

**Comment by bernd** [comment]
@samupr We need the full Graylog server log to be able to debug this. Thank you!

**Comment by jpdstan** [comment]
Hey @bernd we're seeing the same issue. This seems to happen more often when load is heavier (i.e. when the consumption rate is higher). What we have observed is that our consumer group seems to be stuck in a rebalancing state (whereas it should normally be `Stable`). More specifically, we see the consumer group state flapping between `PreparingRebalance` and `CompletingRebalance` every few seconds. Our consumption rate goes down the drain when this occurs (although it's not completely at 0).

The only way we've been able to remediate this is to press "Stop input" and "Start input" in the Inputs page until the consumer group becomes `Stable` again. This can take up to a few tries, in which it may still flap between the aforementioned rebalancing states.

Here's an example of what the graylog consumer logs look like when it's flapping between `PreparingRebalance` and `CompletingRebalance`:

```
2021-02-10T10:09:07.314-05:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
2021-02-10T10:09:07.354-05:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
2021-02-10T10:09:07.354-05:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
2021-02-10T10:09:07.659-05:00 INFO  [AbstractCoordinator] SyncGroup for group graylog failed due to coordinator rebalance, rejoining the group
2021-02-10T10:09:07.663-05:00 INFO  [AbstractCoordinator] SyncGroup for group graylog failed due to coordinator rebalance, rejoining the group
2021-02-10T10:09:08.088-05:00 INFO  [AbstractCoordinator] SyncGroup for group graylog failed due to coordinator rebalance, rejoining the group
2021-02-10T10:09:08.088-05:00 INFO  [AbstractCoordinator] SyncGroup for group graylog failed due to coordinator rebalance, rejoining the group
2021-02-10T10:09:08.088-05:00 INFO  [AbstractCoordinator] SyncGroup for group graylog failed due to coordinator rebalance, rejoining the group
2021-02-10T10:09:08.088-05:00 INFO  [AbstractCoordinator] SyncGroup for group graylog failed due to coordinator rebalance, rejoining the group
2021-02-10T10:09:08.271-05:00 INFO  [AbstractCoordinator] SyncGroup for group graylog failed due to coordinator rebalance, rejoining the group
2021-02-10T10:09:08.364-05:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
2021-02-10T10:09:08.364-05:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
2021-02-10T10:09:08.364-05:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
2021-02-10T10:09:08.364-05:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
2021-02-10T10:09:08.407-05:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
2021-02-10T10:09:08.407-05:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
```

Some important build info:
- Kafka version 2.4.0
- Java 1.8.0
- Graylog-server 3.3.2
```

**Comment by Torch09** [comment]
Hi,

we've got the same Problem.
This occurs not only on kafka GELF but also with the plaintext input. Probably everything which uses Kafka.

**Our setup:**
3x Graylog Server (v 4.0.2) inc. 3x Mongodb
3x Kafka (v 2.7.0)
8x Elasticsearch (v 7.8.0)
OS: Debian 10

**Graylog Kafka Inputs:**
13x Kafka Inputs with 2 processor threads per Input (running on all nodes)


**Things we observed:** 
The issue isn't directly linked to load in our case. 
After starting to many Kafka Inputs the Inputs become unstable. They'll crash after some time, or instantly after starting even more Inputs -> 11 Inputs run fine a while, but they'll crash after some time. With the start of the 12th input, it crashes instantly.
We've got two consumer groups, only the group which is started last crashes, the first one runs fine.
The only way to get the consumers stable again is to stop the inputs.
As @jpdstan pointed out, the kafka version on Graylog-Server seems quite old: `[AppInfoParser] Kafka version : 0.9.0.1`

_________________

**Log snippet:**

[...]
  2021-02-11T12:48:45.060+01:00 INFO  [AbstractCoordinator] SyncGroup for group graylogtraps failed due to coordinator rebalance, rejoining the group
  2021-02-11T12:48:45.062+01:00 INFO  [AbstractCoordinator] SyncGroup for group graylogtraps failed due to coordinator rebalance, rejoining the group
  2021-02-11T12:48:46.546+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:46.546+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:46.550+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:46.550+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:46.589+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:46.589+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:47.059+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:47.059+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:47.061+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:47.061+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:47.491+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:47.491+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:47.601+01:00 INFO  [AbstractCoordinator] Attempt to heart beat failed since the group is rebalancing, try to re-join group.
  2021-02-11T12:48:47.601+01:00 INFO  [AbstractCoordinator] Attempt to heart beat failed since the group is rebalancing, try to re-join group.
  2021-02-11T12:48:47.601+01:00 INFO  [AbstractCoordinator] Attempt to heart beat failed since the group is rebalancing, try to re-join group.
  2021-02-11T12:48:47.601+01:00 INFO  [AbstractCoordinator] Attempt to heart beat failed since the group is rebalancing, try to re-join group.
  2021-02-11T12:48:47.918+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:47.918+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:47.966+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:47.966+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:49.060+01:00 INFO  [AbstractCoordinator] SyncGroup for group graylogtraps failed due to coordinator rebalance, rejoining the group
  2021-02-11T12:48:49.063+01:00 INFO  [AbstractCoordinator] SyncGroup for group graylogtraps failed due to coordinator rebalance, rejoining the group
  2021-02-11T12:48:49.492+01:00 INFO  [AbstractCoordinator] SyncGroup for group graylogtraps failed due to coordinator rebalance, rejoining the group
  2021-02-11T12:48:50.547+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:50.547+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:50.550+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:50.550+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:50.589+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:50.589+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:51.060+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:51.060+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:51.061+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:51.061+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:51.491+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:51.491+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:51.598+01:00 INFO  [AbstractCoordinator] Attempt to heart beat failed since the group is rebalancing, try to re-join group.
  2021-02-11T12:48:51.598+01:00 INFO  [AbstractCoordinator] Attempt to heart beat failed since the group is rebalancing, try to re-join group.
  2021-02-11T12:48:51.598+01:00 INFO  [AbstractCoordinator] Attempt to heart beat failed since the group is rebalancing, try to re-join group.
  2021-02-11T12:48:51.600+01:00 INFO  [AbstractCoordinator] Attempt to heart beat failed since the group is rebalancing, try to re-join group.
  2021-02-11T12:48:51.919+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
  2021-02-11T12:48:51.919+01:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
  2021-02-11T12:48:51.967+01:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
[...]

_________________
 Hope this helps.

**Comment by jpdstan** [comment]
@Torch09 one thing that seemed to help us in the last 24 hours was increasing the default 3s heartbeat interval:
```
custom_properties: heartbeat.interval.ms=10000
```
which might indicate that consumers are being incorrectly marked as dead by the brokers

**Comment by Torch09** [comment]
@jpdstan that doesn't seem to work in our case. 
But thanks for the tip!

**Comment by jpdstan** [comment]
Yeah on our end that seemed to have just delayed the issue - we're still noticing the same rebalancing behavior in the long run.

@bernd any ideas?

**Comment by Torch09** [comment]
We've changed the "global" parameter on the Kafka inputs to specific nodes.
Now it seems to work as a workaround.
It looks like Graylog has problems if it has to handle multiple Kafka inputs per node.

This fixed it for the moment, but no loadbalancing or ha isn't my preferred long time setup 😁

**Comment by bernd** [comment]
Reopen again. @mpfz0r is working on an update to the 2.7.0 version of the Kafka libraries in https://github.com/Graylog2/graylog2-server/pull/10146.

**Comment by lotusnoir** [comment]
Hello,
I'm having the exactly same issue... :/

My setup is:

- 2x graylog 4.0.5-1
- 9x elasticsearch 7.12.0
- 2x mongodb 4.4.4
- 3x kafka 2.7.0
- 3x zookeeper 3.6.2

the inputs on graylog are gelf kafka, i thought it was cause of the cluster mode but even if 1 graylog node is running it's the same, dont find how to make it works, please advise...

```
2021-04-08T09:14:36.094+02:00 INFO  [AbstractCoordinator] SyncGroup for group graylog4 failed due to coordinator rebalance, rejoining the group
2021-04-08T09:14:36.094+02:00 INFO  [AbstractCoordinator] SyncGroup for group graylog4 failed due to coordinator rebalance, rejoining the group
2021-04-08T09:14:36.336+02:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
2021-04-08T09:14:36.336+02:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
2021-04-08T09:14:36.381+02:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
2021-04-08T09:14:36.382+02:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
2021-04-08T09:14:36.761+02:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
2021-04-08T09:14:36.761+02:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
2021-04-08T09:14:36.761+02:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
2021-04-08T09:14:37.034+02:00 INFO  [AbstractCoordinator] SyncGroup for group graylog4 failed due to coordinator rebalance, rejoining the group
2021-04-08T09:14:37.066+02:00 ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
2021-04-08T09:14:37.067+02:00 ERROR [KafkaTransport] Caught recoverable exception. Retrying
```
Thanks in advance

**Comment by mpfz0r** [comment]
@lotusnoir please try `4.0.6` which received an update to our kafka client:
https://www.graylog.org/post/announcing-graylog-v4-0-6

**Comment by lotusnoir** [comment]
indeed !
i made the upgrade to graylog 4.0.6 4h ago and no problem since. I keep watching but it looks like its fixed :)
Thanks a lot !

**Comment by samupr** [comment]
I have updated to v4.0.6 and have noticed that my GELF Kafka Input now isn't load-balancing across my Graylog cluster of 3 servers. 
All messages are getting processed now by a single node. I'm almost sure that in previous version when using v4.0.5 it seemed to load-balance correctly across the cluster.

**Comment by mpfz0r** [comment]
@samupr how many partitions are in your Kafka topic?
Kafka can only share the load with one consumer per partition.
So you might wanna increase your partitions.
It might also help to reduce the number of Processor Threads to 1 on the Kafka inputs.
This way you'd only have one consumer per graylog node.

**Comment by lotusnoir** [comment]
I confirm that on 4.0.6, the load-balancing on Gelf kafka is working properly, it must be a config problem like @mpfz0r says

**Comment by samupr** [comment]
My mistake, I didn't divide the number of partitions in the topic by the 3 Graylog nodes. I confirm that all looks good now. Thanks for your help.

**Comment by Sohan2infoSec** [comment]
Hi team,
I have this issue with Kafka-0.9.0.1

If we use graylog:4.0.5 the log kept emiting
>ERROR [KafkaTransport] Caught recoverable exception. Retrying
ERROR [KafkaTransport] Caught recoverable exception. Retrying
ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.
ERROR [KafkaTransport] WakeupException in poll. Kafka server is not responding.

if we used Graylog:4.0.6 and the log is like below
> 2021-05-04 17:02:50,569 INFO : org.apache.kafka.common.utils.AppInfoParser - Kafka version: 2.7.0
2021-05-04 17:02:50,570 INFO : org.apache.kafka.common.utils.AppInfoParser - Kafka commitId: 448719dc99a19793
2021-05-04 17:02:50,570 INFO : org.apache.kafka.common.utils.AppInfoParser - Kafka startTimeMs: 1620147770569
2021-05-04 17:02:50,570 INFO : org.apache.kafka.clients.consumer.KafkaConsumer - [Consumer clientId=gl2-782687cf-60917d9716851d4a570e36d6-1, groupId=graylog2] Subscribed to pattern: '^f1-ops$'
2021-05-04 17:02:50,575 INFO : org.graylog2.inputs.InputStateListener - Input [Raw/Plaintext Kafka/60917d9716851d4a570e36d6] is now RUNNING
2021-05-04 17:03:21,617 WARN : org.apache.kafka.clients.NetworkClient - [Consumer clientId=gl2-782687cf-60917d9716851d4a570e36d6-1, groupId=graylog2] Bootstrap broker kafka-ip:9092 (id: -1 rack: null) disconnected
2021-05-04 17:03:21,617 WARN : org.apache.kafka.clients.NetworkClient - [Consumer clientId=gl2-782687cf-60917d9716851d4a570e36d6-0, groupId=graylog2] Bootstrap broker kafka-ip:9092 (id: -1 rack: null) disconnected

I don't know it parsed the kafka-version is 2.7.0 although I used 0.9.0.1. Really appreciated if any one have advice for me.
